lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/BuildCriOauthRequestHandler.java

```diff
@@ -333,8 +351,45 @@ private JWEObject signEncryptJar(
                         evidenceRequest,
                         context);
 
-        RSAEncrypter rsaEncrypter = new RSAEncrypter(oauthCriConfig.getParsedEncryptionKey());
-        return AuthorizationRequestHelper.createJweObject(rsaEncrypter, signedJWT);
+        var encKey = getCriEncryptionKey(oauthCriConfig);
+        RSAEncrypter rsaEncrypter = new RSAEncrypter(encKey);
+        return AuthorizationRequestHelper.createJweObject(
+                rsaEncrypter, signedJWT, encKey.getKeyID());
+    }
+
+    private RSAKey getCriEncryptionKey(OauthCriConfig criConfig) throws ParseException {
+        var jwksUrl = criConfig.getJwksUrl();
+
+        if (jwksUrl != null) {
+            var cacheDuration =
+                    configService.getEnvironmentVariable(OAUTH_CRI_ENC_KEY_CACHE_DURATION_MINUTES)
+                                    == null
+                            ? DEFAULT_OAUTH_CRI_ENC_KEY_CACHE_DURATION_MINUTES
+                            : Integer.parseInt(
+                                    configService.getEnvironmentVariable(
+                                            OAUTH_CRI_ENC_KEY_CACHE_DURATION_MINUTES));
+
+            if (cachedOAuthCriEncryptionKey != null
+                    && !cachedOAuthCriEncryptionKey.isExpired(cacheDuration)) {
+                return cachedOAuthCriEncryptionKey.key();
+            }
+
+            var key = criApiService.getKeyFromJwksEndpoint(jwksUrl);
+
+            if (key.isPresent()) {
+                // Cache the key in a variable outside of the handler so it exists in
+                // the Lambda's memory and is still accessible between invocations that
+                // occur at short intervals from one another
+                var parsedKey = key.get().toRSAKey();
+                cacheEncryptionKey(parsedKey);
+                return parsedKey;
+            }
+        }
+        return criConfig.getParsedEncryptionKey();
+    }
+
+    private static void cacheEncryptionKey(RSAKey key) {
+        cachedOAuthCriEncryptionKey = new CachedOAuthCriEncryptionKey(key, new Date());
```

Comment:
(this could just be inlined)

Reply:
it was initially inlined but sonar complained that you shouldn't be mutating a static variable inside a non-static method so I've extracted it here to shush it